scope/nn.py

```diff
@@ -0,0 +1,340 @@
+from abc import ABC
+import datetime
+import os
+import pathlib
+import tensorflow as tf
+from typing import Optional
+
+from .models import AbstractClassifier
+
+
+class DenseBlock(tf.keras.models.Model, ABC):
+    def __init__(
+        self, units: int, activation: str = "relu", repetitions: int = 1, **kwargs
+    ):
+        """Dense block to process light curve features
+
+        :param units: tuple of size 2 with number of units in repeated dense blocks
+        :param activation:
+        :param repetitions:
+        """
+        super(DenseBlock, self).__init__(**kwargs)
+        self.units = units
+        self.repetitions = repetitions
+        self.activation = activation
+
+        for i in range(self.repetitions):
+            vars(self)[f"dense_{i}"] = tf.keras.layers.Dense(
+                units=self.units, activation=self.activation
+            )
+
+    def call(self, inputs, **kwargs):
+        x = self.dense_0(inputs)
+        for i in range(1, self.repetitions):
+            x = vars(self).get(f"dense_{i}")(x)
+
+        return x
+
+
+class ConvBlock(tf.keras.models.Model, ABC):
+    def __init__(
+        self,
+        filters: int,
+        kernel_size: tuple,
+        activation: str = "relu",
+        pool_size: tuple = (2, 2),
+        repetitions: int = 1,
+        **kwargs,
+    ):
+        """Convolutional block to process dmdt's constructed from light curves
+
+        :param filters:
+        :param kernel_size:
+        :param activation:
+        :param pool_size:
+        :param repetitions:
+        :param kwargs:
+        """
+        super(ConvBlock, self).__init__(**kwargs)
+
+        self.filters = filters
+        self.kernel_size = kernel_size
+        self.activation = activation
+        self.pool_size = pool_size
+        self.repetitions = repetitions
+
+        for i in range(self.repetitions):
+            vars(self)[f"conv_{i}"] = tf.keras.layers.SeparableConv2D(
+                filters=self.filters,
+                kernel_size=self.kernel_size,
+                activation=self.activation,
+            )
+
+        self.max_pool = tf.keras.layers.MaxPooling2D(pool_size=self.pool_size)
+
+    def call(self, inputs, **kwargs):
+        x = self.conv_0(inputs)
+        for i in range(1, self.repetitions):
+            x = vars(self).get(f"conv_{i}")(x)
+        x = self.max_pool(x)
+
+        return x
+
+
+class ScopeNet(tf.keras.models.Model, ABC):
```

Comment:
This looks at least similar to paper I, so maybe reference that paper for the network discussion?

Reply:
Good call, added. FYI It's the same architecture, I just significantly improved the code quality - no more spaghetti.